File: PyTorch/sparseconvnet/legacy/affineReLUTrivialConvolution.py
# Copyright 2016-present, Facebook, Inc.
# All rights reserved.
#
# This source code is licensed under the license found in the
# LICENSE file in the root directory of this source tree.

"""
  Affine transformation (i.e. the second half of a typical batchnormalization layer)
  Parameters:
  nPlanes : number of input planes
  noise : add multiplicative and additive noise during training if >0.
  leakiness : Apply activation function inplace: 0<=leakiness<=1.
  0 for ReLU, values in (0,1) for LeakyReLU, 1 for no activation function.
"""

import torch
import sparseconvnet as s
from . import SparseModule
from ..utils import toLongTensor, typed_fn, optionalTensor, nullptr
from .sparseConvNetTensor import SparseConvNetTensor
import math


class AffineReLUTrivialConvolution(SparseModule):
    def __init__(self, nIn, nOut, additiveGrad=False):
        SparseModule.__init__(self)
        self.nIn = nIn
        self.nOut = nOut
        self.affineWeight = torch.Tensor(nIn).fill_(1)
        self.affineBias = torch.Tensor(nIn).zero_()
        self.convWeight = torch.Tensor(
            nIn, nOut).normal_(
            0, math.sqrt(
                2.0 / nIn))
        self.gradAffineWeight = torch.Tensor(nIn).fill_(0)
        self.gradAffineBias = torch.Tensor(nIn).zero_()
        self.gradConvWeight = torch.Tensor(nIn, nOut).zero_()
        self.additiveGrad = additiveGrad
        self.output = SparseConvNetTensor(torch.Tensor())
        self.gradInput = torch.Tensor()

    def parameters(self):
        return [self.affineWeight, self.affineBias, self.convWeight], [
            self.gradAffineWeight, self.gradAffineBias, self.gradConvWeight]

    def updateOutput(self, input):
        self.output.metadata = input.metadata
        self.output.spatial_size = input.spatial_size
        typed_fn(input, 'AffineReluTrivialConvolution_updateOutput')(
            input.features,
            self.output.features,
            self.affineWeight,
            self.affineBias,
            self.convWeight)
        s.forward_pass_multiplyAdd_count += input.features.size(
            0) * self.nIn * self.nOut
        s.forward_pass_hidden_states += self.output.features.nelement()
        return self.output

    def backward(self, input, gradOutput, scale=1):
        assert scale == 1
        typed_fn(input, 'AffineReluTrivialConvolution_backward')(
            input.features,
            self.gradInput,
            gradOutput,
            self.affineWeight,
            self.gradAffineWeight,
            self.affineBias,
            self.gradAffineBias,
            self.convWeight,
            self.gradConvWeight,
            self.additiveGrad)
        return self.gradInput

    def updateGradInput(self, input, gradOutput):
        assert false  # just call backward

    def accGradParameters(input, gradOutput, scale):
        assert false  # just call backward

    def __repr__(self):
        s = 'AffineReluTrivialConvolution ' + \
            str(self.nIn) + '->' + str(self.nOut)
        return s

    def type(self, t=None, tensorCache=None):
        if t is None:
            return self._type
        self._type = t
        self.affineWeight = self.affineWeight.type(t)
        self.affineBias = self.affineBias.type(t)
        self.convWeight = self.convWeight.type(t)
        self.gradAffineWeight = self.gradAffineWeight.type(t)
        self.gradAffineBias = self.gradAffineBias.type(t)
        self.gradConvWeight = self.gradConvWeight.type(t)
        self.gradInput = self.gradInput.type(t)
        self.output.type(t)

File: Torch/AffineReluTrivialConvolution.lua
-- Copyright 2016-present, Facebook, Inc.
-- All rights reserved.
--
-- This source code is licensed under the license found in the
-- LICENSE file in the root directory of this source tree.

return function(sparseconvnet)
  local C = sparseconvnet.C
  local AF,parent = torch.class(
    'sparseconvnet.AffineReluTrivialConvolution', 'nn.Module',sparseconvnet)

  function AF:__init(nInputPlanes, nOutputPlanes,additiveGrad)
    parent.__init(self)
    self.nInputPlanes=nInputPlanes
    self.nOutputPlanes=nOutputPlanes
    self.affineWeight = torch.Tensor(nInputPlanes)
    self.affineBias = torch.Tensor(nInputPlanes)
    self.convWeight = torch.Tensor(nInputPlanes,nOutputPlanes)
    self.gradAffineWeight = torch.Tensor(nInputPlanes):fill(0)
    self.gradAffineBias = torch.Tensor(nInputPlanes):zero()
    self.gradConvWeight = torch.Tensor(nInputPlanes,nOutputPlanes):zero()
    self.additiveGrad=additiveGrad or false --boolean
    self.output={
      features=torch.Tensor(),
    }
    self:reset()
  end

  function AF:reset()
    self.affineWeight:fill(1)
    self.affineBias:zero()
    self.convWeight:normal(0,math.sqrt(2/self.nInputPlanes)) --not 2/self.nOutputPlanes?
  end

  function AF:parameters()
    return {self.affineWeight, self.affineBias, self.convWeight},
    {self.gradAffineWeight, self.gradAffineBias, self.gradConvWeight}
  end

  function AF:updateOutput(input)
    self.output.metadata = input.metadata
    self.output.spatialSize = input.spatialSize
    C.typedFn(self._type,'AffineReluTrivialConvolution_updateOutput')(
      input.features:cdata(),
      self.output.features:cdata(),
      self.affineWeight:cdata(),
      self.affineBias:cdata(),
      self.convWeight:cdata())
    self.shared.forwardPassMultiplyAddCount=
    self.shared.forwardPassMultiplyAddCount+
    input.features:size(1)*self.nInputPlanes*self.nOutputPlanes
    self.shared.forwardPassHiddenStates=
    self.shared.forwardPassHiddenStates+self.output.features:nElement()
    return self.output
  end

  function AF:backward(input, gradOutput)
    C.typedFn(self._type,'AffineReluTrivialConvolution_backward')(
      input.features:cdata(),
      self.gradInput.features:cdata(),
      gradOutput.features:cdata(),
      self.affineWeight:cdata(),
      self.gradAffineWeight:cdata(),
      self.affineBias:cdata(),
      self.gradAffineBias:cdata(),
      self.convWeight:cdata(),
      self.gradConvWeight:cdata(),
      self.additiveGrad)
    return self.gradInput
  end

  function AF:updateGradInput(input, gradOutput)
    assert(false) --just call backward
  end

  function AF:accGradParameters(input, gradOutput, scale)
    assert(false) --just call backward
  end

  function AF:__tostring()
    local s = 'AffineReluTrivialConvolution(' ..
    self.nInputPlanes..'->' .. self.nOutputPlanes .. ')'
    return s
  end

  function AF:clearState()
    self.output={features=self.output.features:set()}
    self.gradInput={features=self.gradInput.features:set()}
    self.rules=nil
  end

  function AF:suggestInputSize(nOut)
    return nOut
  end
end
